plugins/auth-backend/src/providers/azure/provider.ts

```diff
@@ -0,0 +1,290 @@
+import { Config } from '@backstage/config';
+import passport from 'passport';
+
+import got from 'got';
+
+type PrivateInfo = {
+  refreshToken: string;
+};
+
+export class AzureAuthProvider implements OAuthProviderHandlers {
+  private readonly _strategy: AzureStrategy;
+
+  static transformAuthResponse(
+    rawProfile: any,
+    accessToken: string,
+    params: any,
+    photoURL?: any,
+  ): OAuthResponse {
+    const passportProfile: passport.Profile = {
+      id: rawProfile.oid,
+      username: rawProfile._json.preferred_username,
+      displayName: rawProfile.displayName,
+      provider: 'azure',
+      name: {
+        familyName: rawProfile.name.familyName,
+        givenName: rawProfile.name.givenName,
+      },
+      // If no email for user, fallback to preferred_username
+      emails: [
+        {
+          value: rawProfile._json.email || rawProfile._json.preferred_username,
+        },
+      ],
+      photos: [{ value: photoURL }],
+    };
+
+    const profile = makeProfileInfo(passportProfile, params.id_token);
+
+    const providerInfo = {
+      idToken: params.id_token,
+      accessToken,
+      scope: params.scope,
+      expiresInSeconds: Number(params.expires_in),
+    };
+
+    return {
+      providerInfo,
+      profile,
+    };
+  }
+
+  constructor(options: AzureProviderOptions) {
+    this._strategy = new AzureStrategy(
+      { ...options },
+      (
+        _req,
+        _sub,
+        _iss,
+        rawProfile,
+        _jwtClaims,
+        accessToken,
+        refreshToken,
+        params,
+        done: PassportDoneCallback<OAuthResponse, PrivateInfo>,
+      ) => {
+        got
+          .get('https://graph.microsoft.com/v1.0/me/photos/48x48/$value', {
+            encoding: 'binary',
+            responseType: 'buffer',
+            headers: {
+              Authorization: `Bearer ${accessToken}`,
+            },
+          })
+          .then(photoData => {
+            const photoURL = `data:image/jpeg;base64,${Buffer.from(
+              photoData.body,
+            ).toString('base64')}`;
+            const authResponse = AzureAuthProvider.transformAuthResponse(
+              rawProfile,
+              accessToken,
+              params,
+              photoURL,
+            );
+
+            done(undefined, authResponse, { refreshToken });
+          })
+          .catch(error => {
+            console.log(
+              `Error retrieving user photo from Microsoft Graph API: ${error}`,
+            );
+            const authResponse = AzureAuthProvider.transformAuthResponse(
+              rawProfile,
+              accessToken,
+              params,
+            );
+
+            done(undefined, authResponse, { refreshToken });
+          });
+      },
+    );
+  }
+
+  async start(
+    req: express.Request,
+    options: Record<string, string>,
+  ): Promise<RedirectInfo> {
+    const providerOptions = {
+      ...options,
+      prompt: 'consent',
+      customState: options.state,
```

Comment:
This passing the encoded `state` value into the `customState` parameter of `passport-azure-ad`.